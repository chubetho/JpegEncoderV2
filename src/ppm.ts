import * as fs from 'node:fs'

export function readPpm(path: string) {
  const content = fs.readFileSync(path, 'ascii')
  const lines = content.split('\n')
  const [format, size, _maxColor, ...rest] = lines
  const [imageWidth, imageHeight] = size.trim().split(' ').map(Number)
  const maxColor = Number.parseInt(_maxColor)

  if (format !== 'P3')
    throw new Error ('Invalid ppm')

  if (maxColor !== 255)
    throw new Error ('Invalid ppm')

  const image = new Uint8Array(imageWidth * imageHeight * 3)
  let index = 0

  for (let l = 0; l < rest.length; l++) {
    const trim = rest[l].trim()
    if (trim[0] === '#')
      continue

    const row = trim.split(' ')
    for (let i = 0; i < row.length; i++) {
      if (!row[i])
        continue
      const v = Number.parseInt(row[i])
      if (Number.isNaN(v) || v < 0 || v > maxColor)
        throw new Error ('Invalid ppm')

      image[index++] = v
    }
  }

  const blockHeight = ~~((imageHeight + 7) / 8)
  const blockWidth = ~~((imageWidth + 7) / 8)
  const blocks = Array.from({ length: blockHeight * blockWidth }, () => ({
    R: new Uint8Array(64),
    G: new Uint8Array(64),
    B: new Uint8Array(64),
    Y: new Int32Array(64),
    Cb: new Int32Array(64),
    Cr: new Int32Array(64),
  }))

  const clamp = (x: number) => x < -128 ? -128 : x > 127 ? 127 : x
  for (let h = 0; h < imageHeight; h++) {
    const blockRow = ~~(h / 8)
    const pixelRow = h % 8
    for (let w = 0; w < imageWidth; w++) {
      const blockColumn = ~~(w / 8)
      const pixelColumn = w % 8
      const blockIndex = blockRow * blockWidth + blockColumn
      const pixelIndex = pixelRow * 8 + pixelColumn

      const index = (h * imageWidth + w) * 3
      const r = image[index]
      const g = image[index + 1]
      const b = image[index + 2]

      blocks[blockIndex].R[pixelIndex] = r
      blocks[blockIndex].G[pixelIndex] = g
      blocks[blockIndex].B[pixelIndex] = b

      const y = 0.299 * r + 0.587 * g + 0.114 * b - 128
      const cb = -0.1687 * r + -0.3312 * g + 0.5 * b
      const cr = 0.5 * r + -0.4186 * g + -0.0813 * b

      blocks[blockIndex].Y[pixelIndex] = clamp(y)
      blocks[blockIndex].Cb[pixelIndex] = clamp(cb)
      blocks[blockIndex].Cr[pixelIndex] = clamp(cr)
    }
  }

  return {
    blocks,
    metadata: {
      imageHeight,
      imageWidth,
      format,
      maxColor,
      blockHeight,
      blockWidth,
    },
  }
}

export function writePpm(path: string, { blocks, metadata }: Image) {
  let data = `${metadata.format}\n${metadata.imageWidth} ${metadata.imageHeight}\n${metadata.maxColor}\n`

  for (let h = 0; h < metadata.imageHeight; h++) {
    const blockRow = ~~(h / 8)
    const pixelRow = h % 8
    for (let w = 0; w < metadata.imageWidth; w++) {
      const blockColumn = ~~(w / 8)
      const pixelColumn = w % 8
      const blockIndex = blockRow * metadata.blockWidth + blockColumn
      const pixelIndex = pixelRow * 8 + pixelColumn

      const r = blocks[blockIndex].R[pixelIndex]
      const g = blocks[blockIndex].G[pixelIndex]
      const b = blocks[blockIndex].B[pixelIndex]
      data += `${r} ${g} ${b}  `
    }

    data += '\n'
  }

  fs.writeFileSync(path, data)
}
