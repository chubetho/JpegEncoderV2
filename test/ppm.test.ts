import { expect, it } from 'bun:test'
import { readPpm, writePpm } from '../src/ppm.js'

it('read ppm', () => {
  const image = readPpm('src/assets/img_1_32.ppm')
  expect(image).toMatchSnapshot()
})

it.skip('write ppm', () => {
  const image = readPpm('src/assets/img_1_1.ppm')
  writePpm('src/output/write.ppm', image)
})

it('read big ppm', () => {
  readPpm('src/assets/big.ppm')
})
