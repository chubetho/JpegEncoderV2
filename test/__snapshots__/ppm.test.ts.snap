// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`read ppm 1`] = `
{
  "blocks": [
    {
      "Cb": Int16Array [
        -4,
        2,
        -3,
        2,
        -10,
        -12,
        -5,
        -22,
        13,
        1,
        -12,
        2,
        -1,
        -20,
        4,
        -19,
        -5,
        -5,
        -14,
        8,
        4,
        -20,
        2,
        -48,
        -12,
        1,
        -8,
        5,
        -2,
        6,
        -14,
        -46,
        -16,
        7,
        6,
        -1,
        -1,
        -20,
        -21,
        -26,
        -11,
        0,
        8,
        2,
        -2,
        -15,
        -13,
        -9,
        -4,
        -3,
        2,
        0,
        -4,
        1,
        -9,
        -24,
        -5,
        -3,
        0,
        -2,
        -6,
        3,
        -22,
        -21,
      ],
      "Cr": Int16Array [
        -1,
        0,
        6,
        -2,
        5,
        8,
        4,
        19,
        1,
        1,
        2,
        2,
        0,
        10,
        2,
        23,
        2,
        4,
        7,
        2,
        0,
        15,
        13,
        26,
        2,
        1,
        1,
        -1,
        0,
        1,
        15,
        21,
        10,
        8,
        0,
        0,
        0,
        13,
        14,
        9,
        8,
        3,
        -4,
        0,
        -2,
        6,
        7,
        4,
        3,
        6,
        0,
        0,
        3,
        5,
        12,
        9,
        4,
        1,
        1,
        1,
        15,
        14,
        12,
        5,
      ],
      "Y": Int16Array [
        -110,
        -111,
        -105,
        -114,
        -94,
        -80,
        -81,
        -53,
        -109,
        -115,
        -92,
        -102,
        -103,
        -71,
        -94,
        -66,
        -88,
        -83,
        -68,
        -96,
        -109,
        -77,
        -92,
        -21,
        -67,
        -79,
        -76,
        -94,
        -95,
        -111,
        -85,
        -23,
        -53,
        -84,
        -107,
        -103,
        -103,
        -70,
        -54,
        -34,
        -76,
        -90,
        -115,
        -115,
        -108,
        -80,
        -69,
        -68,
        -98,
        -99,
        -108,
        -109,
        -103,
        -109,
        -91,
        -57,
        -88,
        -91,
        -99,
        -100,
        -94,
        -110,
        -72,
        -48,
      ],
    },
    {
      "Cb": Int16Array [
        -18,
        -22,
        -6,
        -7,
        -3,
        -1,
        -11,
        -1,
        -23,
        4,
        -21,
        -10,
        4,
        -3,
        -6,
        -24,
        10,
        16,
        -14,
        4,
        2,
        0,
        -4,
        -27,
        18,
        9,
        -2,
        -4,
        -2,
        -5,
        0,
        -16,
        5,
        4,
        -9,
        -2,
        -9,
        -19,
        -1,
        -3,
        -6,
        -15,
        -15,
        -11,
        -9,
        -8,
        0,
        -5,
        -16,
        -35,
        -10,
        -7,
        -10,
        0,
        2,
        -7,
        -21,
        -4,
        3,
        0,
        1,
        -10,
        -9,
        -18,
      ],
      "Cr": Int16Array [
        20,
        23,
        13,
        7,
        4,
        0,
        1,
        13,
        13,
        11,
        28,
        14,
        0,
        1,
        3,
        8,
        6,
        -10,
        10,
        1,
        -1,
        -1,
        3,
        2,
        11,
        0,
        -1,
        1,
        0,
        1,
        3,
        4,
        1,
        7,
        5,
        4,
        1,
        4,
        2,
        5,
        9,
        18,
        18,
        9,
        5,
        3,
        2,
        5,
        23,
        35,
        14,
        11,
        10,
        0,
        0,
        8,
        20,
        18,
        4,
        -3,
        5,
        16,
        6,
        11,
      ],
      "Y": Int16Array [
        -50,
        -42,
        -70,
        -82,
        -92,
        -99,
        -83,
        -82,
        -57,
        -83,
        -65,
        -83,
        -114,
        -107,
        -97,
        -60,
        -72,
        -105,
        -79,
        -103,
        -112,
        -113,
        -104,
        -62,
        -81,
        -109,
        -106,
        -100,
        -99,
        -95,
        -99,
        -78,
        -61,
        -81,
        -79,
        -86,
        -80,
        -55,
        -70,
        -80,
        -68,
        -58,
        -63,
        -71,
        -71,
        -70,
        -81,
        -85,
        -42,
        -27,
        -67,
        -78,
        -73,
        -92,
        -99,
        -90,
        -26,
        -57,
        -89,
        -93,
        -95,
        -85,
        -92,
        -70,
      ],
    },
    {
      "Cb": Int16Array [
        -7,
        -17,
        -28,
        1,
        0,
        0,
        0,
        0,
        -11,
        -19,
        -1,
        7,
        0,
        0,
        0,
        0,
        -12,
        2,
        -5,
        -3,
        0,
        0,
        0,
        0,
        -20,
        -3,
        -6,
        -8,
        0,
        0,
        0,
        0,
        -36,
        -2,
        0,
        5,
        0,
        0,
        0,
        0,
        -30,
        -10,
        22,
        0,
        0,
        0,
        0,
        0,
        -27,
        -4,
        16,
        -9,
        0,
        0,
        0,
        0,
        -9,
        13,
        -14,
        -6,
        0,
        0,
        0,
        0,
      ],
      "Cr": Int16Array [
        14,
        12,
        16,
        7,
        0,
        0,
        0,
        0,
        7,
        6,
        2,
        0,
        0,
        0,
        0,
        0,
        0,
        2,
        2,
        1,
        0,
        0,
        0,
        0,
        1,
        2,
        4,
        6,
        0,
        0,
        0,
        0,
        13,
        3,
        1,
        0,
        0,
        0,
        0,
        0,
        6,
        4,
        -2,
        0,
        0,
        0,
        0,
        0,
        8,
        0,
        0,
        4,
        0,
        0,
        0,
        0,
        9,
        -1,
        6,
        3,
        0,
        0,
        0,
        0,
      ],
      "Y": Int16Array [
        -85,
        -64,
        -36,
        -67,
        0,
        0,
        0,
        0,
        -54,
        -37,
        -56,
        -82,
        0,
        0,
        0,
        0,
        -49,
        -63,
        -71,
        -74,
        0,
        0,
        0,
        0,
        -51,
        -55,
        -61,
        -65,
        0,
        0,
        0,
        0,
        -29,
        -42,
        -61,
        -89,
        0,
        0,
        0,
        0,
        -44,
        -38,
        -91,
        -103,
        0,
        0,
        0,
        0,
        -54,
        -62,
        -98,
        -89,
        0,
        0,
        0,
        0,
        -72,
        -110,
        -87,
        -86,
        0,
        0,
        0,
        0,
      ],
    },
    {
      "Cb": Int16Array [
        -9,
        -2,
        5,
        2,
        0,
        -6,
        0,
        -1,
        -13,
        -5,
        7,
        -4,
        -3,
        -6,
        1,
        -6,
        -4,
        -3,
        -7,
        -10,
        2,
        1,
        1,
        -5,
        -10,
        7,
        -40,
        -7,
        10,
        -28,
        4,
        9,
        -10,
        -1,
        -15,
        -17,
        2,
        -3,
        -10,
        6,
        -3,
        -34,
        11,
        -16,
        11,
        6,
        -11,
        2,
        -13,
        -33,
        9,
        6,
        6,
        -3,
        -15,
        -19,
        -18,
        -23,
        5,
        10,
        -3,
        -2,
        -22,
        -32,
      ],
      "Cr": Int16Array [
        5,
        2,
        -1,
        -1,
        1,
        10,
        5,
        0,
        4,
        2,
        -2,
        1,
        0,
        2,
        5,
        4,
        -1,
        -1,
        6,
        6,
        0,
        0,
        0,
        2,
        0,
        -2,
        22,
        13,
        -2,
        16,
        3,
        -3,
        2,
        2,
        5,
        0,
        0,
        8,
        7,
        0,
        -1,
        8,
        -1,
        0,
        -3,
        0,
        4,
        0,
        0,
        5,
        4,
        0,
        -1,
        1,
        2,
        7,
        8,
        3,
        -2,
        0,
        2,
        0,
        6,
        16,
      ],
      "Y": Int16Array [
        -79,
        -87,
        -106,
        -113,
        -112,
        -100,
        -104,
        -105,
        -89,
        -88,
        -110,
        -105,
        -101,
        -92,
        -99,
        -97,
        -107,
        -101,
        -89,
        -86,
        -101,
        -106,
        -113,
        -103,
        -89,
        -104,
        -36,
        -54,
        -95,
        -50,
        -83,
        -113,
        -85,
        -88,
        -70,
        -48,
        -61,
        -66,
        -70,
        -98,
        -100,
        -41,
        -71,
        -51,
        -79,
        -97,
        -85,
        -98,
        -76,
        -27,
        -55,
        -81,
        -101,
        -97,
        -74,
        -49,
        -69,
        -46,
        -66,
        -93,
        -94,
        -94,
        -60,
        -16,
      ],
    },
    {
      "Cb": Int16Array [
        -29,
        -4,
        5,
        1,
        0,
        -23,
        -5,
        -12,
        -22,
        -3,
        4,
        1,
        1,
        -9,
        1,
        -3,
        -27,
        2,
        9,
        -1,
        -1,
        -4,
        -19,
        -11,
        -4,
        -1,
        -9,
        0,
        -16,
        1,
        -5,
        -9,
        -6,
        -4,
        -13,
        -7,
        -13,
        9,
        0,
        2,
        -9,
        -2,
        -1,
        -17,
        3,
        2,
        -9,
        -2,
        -21,
        7,
        7,
        -11,
        3,
        -3,
        5,
        0,
        -22,
        28,
        -10,
        -14,
        12,
        5,
        -4,
        -14,
      ],
      "Cr": Int16Array [
        14,
        10,
        7,
        1,
        -3,
        28,
        9,
        3,
        9,
        5,
        13,
        3,
        -3,
        12,
        -2,
        0,
        7,
        7,
        0,
        5,
        -1,
        0,
        4,
        1,
        0,
        0,
        3,
        -1,
        0,
        0,
        4,
        0,
        1,
        2,
        5,
        0,
        -1,
        -3,
        -2,
        -3,
        1,
        1,
        2,
        4,
        -3,
        1,
        0,
        0,
        13,
        0,
        -3,
        2,
        -2,
        1,
        -2,
        -1,
        11,
        -2,
        0,
        1,
        -3,
        -1,
        0,
        7,
      ],
      "Y": Int16Array [
        -59,
        -70,
        -92,
        -98,
        -96,
        -57,
        -89,
        -85,
        -66,
        -74,
        -91,
        -102,
        -103,
        -88,
        -106,
        -102,
        -59,
        -76,
        -104,
        -108,
        -107,
        -101,
        -71,
        -60,
        -105,
        -104,
        -90,
        -96,
        -73,
        -83,
        -81,
        -77,
        -94,
        -92,
        -76,
        -71,
        -57,
        -81,
        -86,
        -96,
        -86,
        -89,
        -92,
        -71,
        -87,
        -96,
        -86,
        -88,
        -28,
        -71,
        -91,
        -80,
        -95,
        -93,
        -107,
        -110,
        -8,
        -87,
        -77,
        -61,
        -91,
        -108,
        -100,
        -78,
      ],
    },
    {
      "Cb": Int16Array [
        5,
        3,
        -8,
        5,
        0,
        0,
        0,
        0,
        5,
        -2,
        0,
        5,
        0,
        0,
        0,
        0,
        10,
        9,
        -3,
        1,
        0,
        0,
        0,
        0,
        1,
        9,
        1,
        -4,
        0,
        0,
        0,
        0,
        -10,
        3,
        12,
        -8,
        0,
        0,
        0,
        0,
        -16,
        8,
        4,
        -6,
        0,
        0,
        0,
        0,
        -11,
        -11,
        2,
        4,
        0,
        0,
        0,
        0,
        -4,
        -4,
        0,
        5,
        0,
        0,
        0,
        0,
      ],
      "Cr": Int16Array [
        0,
        -2,
        0,
        -1,
        0,
        0,
        0,
        0,
        -3,
        -1,
        -2,
        -3,
        0,
        0,
        0,
        0,
        -3,
        -2,
        0,
        -2,
        0,
        0,
        0,
        0,
        -3,
        -3,
        0,
        0,
        0,
        0,
        0,
        0,
        5,
        0,
        -3,
        1,
        0,
        0,
        0,
        0,
        6,
        -1,
        -2,
        -1,
        0,
        0,
        0,
        0,
        4,
        0,
        -4,
        -2,
        0,
        0,
        0,
        0,
        8,
        2,
        -2,
        -2,
        0,
        0,
        0,
        0,
      ],
      "Y": Int16Array [
        -105,
        -115,
        -100,
        -112,
        0,
        0,
        0,
        0,
        -114,
        -108,
        -109,
        -116,
        0,
        0,
        0,
        0,
        -89,
        -111,
        -107,
        -108,
        0,
        0,
        0,
        0,
        -85,
        -108,
        -111,
        -104,
        0,
        0,
        0,
        0,
        -79,
        -95,
        -121,
        -105,
        0,
        0,
        0,
        0,
        -65,
        -95,
        -108,
        -99,
        0,
        0,
        0,
        0,
        -89,
        -76,
        -87,
        -99,
        0,
        0,
        0,
        0,
        -84,
        -91,
        -98,
        -109,
        0,
        0,
        0,
        0,
      ],
    },
  ],
  "metadata": {
    "blockHeight": 2,
    "blockWidth": 3,
    "format": "P3",
    "imageHeight": 16,
    "imageWidth": 20,
    "maxColor": 255,
  },
}
`;
